---
title: "coordinates"
author: "Ameek Bhalla"
date: "10/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(readxl)
library(tidyverse)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
# path <- "GSE22069 files.xlsx"
# 
# wb <- path %>%
#         excel_sheets() %>%
#         set_names() %>%
#        map(read_excel,
#            path = path, n_max = 100)

sheet_1 <- read_excel("GSE22069 files.xlsx", sheet = 1, 
                      # n_max = 100, 
                      col_types = c("guess", "guess", "guess", "guess", "skip", "skip"))

sheet_2 <- read_excel("GSE22069 files.xlsx", sheet = 2, 
                      # n_max = 100, 
                      skip = 21,
                      col_types = c(rep("guess", 5), rep("skip", 6)))

sheet_1 <- sheet_1 %>% mutate(chromosome = 
                                str_split_fixed(sheet_1$seqname, "chr", 2) %>% as_tibble() %>% pluck(2) %>% as_factor())

sheet_2 <- sheet_2 %>% mutate(chromosome_2 = case_when(
                                 chromosome == "2L" | chromosome == "2LHet" ~ "2L",
                                 chromosome == "2R" | chromosome == "2RHet" ~ "2R",
                                 chromosome == "3L" | chromosome == "3LHet" ~ "3L",
                                 chromosome == "3R" | chromosome == "3RHet" ~ "3R",
                                 chromosome == "4" ~ "4",
                                 chromosome == "X" | chromosome == "XHet" ~ "X",
                                 chromosome == "U" | chromosome == "YHet" ~ "Y",
                                 TRUE ~ "dmel_mitochondrion_genome"
                               ) %>% as_factor())


for (i in 1:length(sheet_1$end)) {
sheet_1$genes[[i]]  <- which(sheet_1$chromosome[i] == sheet_2$chromosome_2 & 
                              sheet_1$end[i] >= sheet_2$start & sheet_2$start >= sheet_1$start[i] )  %>%
    as.list()
  }

```

