---
output: html_document
editor_options: 
  chunk_output_type: console
---

***
# Setup
The default chunk options can be set in the first chunk itself. Later, the behaviour of individuals chunks can be modified by modifying a specific argument.

```{r setup results = "hide"}
## Set default chunk options

knitr::opts_chunk$set(results = "hide", 
                      tidy = TRUE)
## Load Packages
library(affy) # Bioconductor package for analysis of older Affymetrix arrays
#library(affyio)
#library(affyPLM)
#library(annaffy)
#library(annotate)
#library(arrayQualityMetrics)
library(BiocManager)
library(drosophila2cdf)
# library(formatR)
library(gcrma)
library(limma)
#library(simpleaffy)
library(tidyverse)
```

# Importing data into R
```{r get_data}

kp_samples <- readxl::read_excel("F0_kp_samples_arranged_coding.xlsx", sheet = "All three generations ")

kp_samples <- kp_samples |> dplyr::select(cel_file_name, `sample code`,  
                                          generation, replicate, temperature, genotype) |> 
  mutate(
  generation = generation |> as.factor(),
  replicate = replicate |> as.factor(),
  temperature = temperature |> as.factor(),
  genotype = case_when(
    genotype == "w1118" ~ "control",
    TRUE ~ "test") |> factor(levels = c("control", "test")),
  label = paste(generation, genotype, temperature, sep = "_") |> as.factor()
  )

# regenerating the output of above steps in case the xlsx file is unavailable
# kp_samples <- data.frame(
#   stringsAsFactors = FALSE,
#      cel_file_name = c("@52008700992223031419433679647524","@52008700992223031419433679647597",
#                        "@52008700992223031419433679647562",
#                        "@52008700992223031419433679647593","@52008700992223031419433679647603",
#                        "@52008700992223031419433679647601","@52008700992223031419433679647592",
#                        "@52008700992223031419433679647527",
#                        "@52008700992223031419433679647544","@52008700992223031419433679647535",
#                        "@52008700992223031419433679647575",
#                        "@52008700992223031419433679647594","@52008700992223031419433679647545",
#                        "@52008700992223031419433679647565",
#                        "@52008700992223031419433679647585","@52008700992223031419433679647569",
#                        "@52008700992223031419433679647589_(Drosophila_2)",
#                        "@52008700949546080816423851715032_(Drosophila_2)",
#                        "@52008700992223031419433679647547_(Drosophila_2)",
#                        "@52008700992223031419433679647591_(Drosophila_2)",
#                        "@52008700992223031419433679647598_(Drosophila_2)",
#                        "@52008700992223031419433679647600_(Drosophila_2)",
#                        "@52008700992223031419433679647548_(Drosophila_2)",
#                        "@52008700992223031419433679647571_(Drosophila_2)",
#                        "@52008700992223031419433679647610_(Drosophila_2)","@52008700992223031419433679647572_(Drosophila_2)",
#                        "@52008700992223031419433679647563_(Drosophila_2)",
#                        "@52008700992213031419433679647427_(Drosophila_2)",
#                        "@52008700992223031419433679647584_(Drosophila_2)",
#                        "@52008700992223031419433679647622_(Drosophila_2)",
#                        "@52008700992223031419433679647549_(Drosophila_2)",
#                        "@52008700992223031419433679647621_(Drosophila_2)",
#                        "@52008700992223031419433679647583_(Drosophila_2)",
#                        "@52008700992213031419433679647476_(Drosophila_2)"),
#        sample.code = c("1W","2W","3W","4W","5T",
#                        "6T","7T","8T","9W","10W","11W","12W","13T","14T",
#                        "15T","16T","C1F1","C2F1","C3F1","C4F1","T1F1",
#                        "T2F1","T3F1","T4F1","C1F2","C2F2","C3F2","C4F2",
#                        "C5F2","T1F2","T2F2","T3F2","T4F2","T5F2"),
#         generation = c("F0","F0","F0","F0","F0",
#                        "F0","F0","F0","F0","F0","F0","F0","F0","F0","F0",
#                        "F0","F1","F1","F1","F1","F1","F1","F1","F1",
#                        "F2","F2","F2","F2","F2","F2","F2","F2","F2","F2"),
#          replicate = c(1L,2L,3L,4L,1L,2L,3L,4L,
#                        1L,2L,3L,4L,1L,2L,3L,4L,1L,2L,3L,4L,1L,2L,
#                        3L,4L,1L,2L,3L,4L,5L,1L,2L,3L,4L,5L),
#        temperature = c("25d","25d","25d","25d",
#                        "25d","25d","25d","25d","16d","16d","16d","16d",
#                        "16d","16d","16d","16d","25d","25d","25d","25d","25d",
#                        "25d","25d","25d","25d","25d","25d","25d","25d",
#                        "25d","25d","25d","25d","25d"),
#           genotype = c("control","control","control",
#                        "control","test","test","test","test","control",
#                        "control","control","control","test","test","test",
#                        "test","control","control","control","control","test",
#                        "test","test","test","control","control","control",
#                        "control","control","test","test","test","test",
#                        "test"),
#              label = c("F0_control_25d",
#                        "F0_control_25d","F0_control_25d","F0_control_25d","F0_test_25d",
#                        "F0_test_25d","F0_test_25d","F0_test_25d",
#                        "F0_control_16d","F0_control_16d","F0_control_16d","F0_control_16d",
#                        "F0_test_16d","F0_test_16d","F0_test_16d",
#                        "F0_test_16d","F1_control_25d","F1_control_25d","F1_control_25d",
#                        "F1_control_25d","F1_test_25d","F1_test_25d",
#                        "F1_test_25d","F1_test_25d","F2_control_25d","F2_control_25d",
#                        "F2_control_25d","F2_control_25d","F2_control_25d",
#                        "F2_test_25d","F2_test_25d","F2_test_25d","F2_test_25d",
#                        "F2_test_25d")
# )
```

```{r preprocessing}
raw_all <- ReadAffy(celfile.path = "./kp CEL")

phenoData(raw_all) <- AnnotatedDataFrame(kp_samples)

sampleNames(raw_all) <- paste(phenoData(raw_all)$generation,
                          phenoData(raw_all)$genotype,
                          phenoData(raw_all)$temperature,
                          phenoData(raw_all)$replicate,
                          sep = "_")

validObject(raw_all)

eset_all <- gcrma(raw_all)

remove(affinity.spline.coefs)
```


```{r qc}

library("arrayQualityMetrics")

# arrayQualityMetrics(expressionset = raw_all, # this can be an AffyBacth object
#                     outdir = "Report_for_raw_data", # directory for saving the output file 'index.html'
#                     force = TRUE, #overwrite existing files of output directory
#                     do.logtransform = TRUE, # log transform the raw intensities
#                     intgroup = "label" # the name of the covariate(s) used to colour the plots
#                     ) 
#                     
# one sample in F0 and one is F1 are problematic

######
# PCA on whole data to find problematic samples
exp_raw <- log2(Biobase::exprs(eset_all))
PCA_raw <- prcomp( #Performs PCA on given matrix and returns results as an object of class prcomp
  t(exp_raw), #transpose
  scale. = FALSE
  )

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2), 1) #round to first place after decimal
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    sample_code = pData(eset_all)$`sample code`
                    )

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point() +
  geom_text(aes(label = sample_code)) +
  ggtitle("The problematic samples") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  theme_classic()

remove(exp_raw, PCA_raw, percentVar, sd_ratio, dataGG)

raw_all <- raw_all[ , -(which(kp_samples$`sample code` == "9W" | kp_samples$`sample code` == "T2F1"))]
kp_samples <- kp_samples[-(which(kp_samples$`sample code` == "9W" | kp_samples$`sample code` == "T2F1")), ]


phenoData(raw_all) <- AnnotatedDataFrame(kp_samples)
sampleNames(raw_all) <- paste(kp_samples$generation,
                          kp_samples$genotype,
                          kp_samples$temperature,
                          kp_samples$replicate,
                          sep = "_")

validObject(raw_all)

eset_all <- gcrma(raw_all)

remove(affinity.spline.coefs)
```

```{r splitting}
raw_F0 <- raw_all[ , which(kp_samples$generation == "F0")]
phenoData(raw_F0) <- AnnotatedDataFrame(kp_samples %>% filter(generation == "F0"))
sampleNames(raw_F0) <- paste(phenoData(raw_F0)$generation,
                          phenoData(raw_F0)$genotype,
                          phenoData(raw_F0)$temperature,
                          phenoData(raw_F0)$replicate,
                          sep = "_")

raw_F1 <- raw_all[ , which(kp_samples$generation == "F1")]
phenoData(raw_F1) <- AnnotatedDataFrame(kp_samples %>% filter(generation == "F1"))
sampleNames(raw_F1) <- paste(phenoData(raw_F1)$generation,
                          phenoData(raw_F1)$genotype,
                          phenoData(raw_F1)$temperature,
                          phenoData(raw_F1)$replicate,
                          sep = "_")

raw_F2 <- raw_all[ , which(kp_samples$generation == "F2")]
phenoData(raw_F2) <- AnnotatedDataFrame(kp_samples %>% filter(generation == "F2"))
sampleNames(raw_F2) <- paste(phenoData(raw_F2)$generation,
                          phenoData(raw_F2)$genotype,
                          phenoData(raw_F2)$temperature,
                          phenoData(raw_F2)$replicate,
                          sep = "_")

# one-line check
validObject(raw_F0)
validObject(raw_F1)
validObject(raw_F2)

eset_F0 <- gcrma(raw_F0)
eset_F1 <- gcrma(raw_F1)
eset_F2 <- gcrma(raw_F2)

remove(affinity.spline.coefs)
```

```{r pca_plots}
######
# PCA of whole data
exp_raw <- log2(Biobase::exprs(eset_all))
PCA_raw <- prcomp( #Performs PCA on given matrix and returns results as an object of class prcomp
  t(exp_raw), #transpose
  scale. = FALSE
  )

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2), 1) #round to first place after decimal
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    generation = pData(eset_all)$generation,
                    genotype = pData(eset_all)$genotype
                    )

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(
        aes(shape = genotype, colour = generation)
        ) +
  ggtitle("PCA of whole data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  theme_classic()

remove(exp_raw, PCA_raw, percentVar, sd_ratio, dataGG)

######
# PCA of F0 data
exp_raw <- log2(Biobase::exprs(eset_F0))
PCA_raw <- prcomp( #Performs PCA on given matrix and returns results as an object of class prcomp
  t(exp_raw), #transpose
  scale. = FALSE
  )

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2), 1) #round to first place after decimal
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    genotype = pData(eset_F0)$genotype,
                    temperature = pData(eset_F0)$temperature
                    )

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(
        aes(shape = genotype, colour = temperature)
        ) +
  ggtitle("PCA of F0 data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  theme_classic()

remove(exp_raw, PCA_raw, percentVar, sd_ratio, dataGG)

######
# PCA of F1 data
exp_raw <- log2(Biobase::exprs(eset_F1))
PCA_raw <- prcomp( #Performs PCA on given matrix and returns results as an object of class prcomp
  t(exp_raw), #transpose
  scale. = FALSE
  )

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2), 1) #round to first place after decimal
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    genotype = pData(eset_F1)$genotype
                    )

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = genotype, colour = genotype)) +
  ggtitle("PCA of F1 data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  theme_classic()

remove(exp_raw, PCA_raw, percentVar, sd_ratio, dataGG)

#######
# PCA of F2 data

exp_raw <- log2(Biobase::exprs(eset_F2))
PCA_raw <- prcomp( #Performs PCA on given matrix and returns results as an object of class prcomp
  t(exp_raw), #transpose
  scale. = FALSE
  )

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2), 1) #round to first place after decimal
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    genotype = pData(eset_F2)$genotype
                    )

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = genotype, colour = genotype)) +
  # geom_text(label = dataGG$replicate) + # replicate L9 separates furthest along the x-axis
  ggtitle("PCA of F2 data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  theme_classic()

remove(exp_raw, PCA_raw, percentVar, sd_ratio, dataGG)

```

# Differential expression analysis
```{r f0_results}
#####

pData(eset_F0) <- pData(eset_F0) |> droplevels()

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 + label, data = pData(eset_F0))

colnames(design_1) <- levels(pData(eset_F0)$label)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_F0, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_F0 <- makeContrasts(
  F0_t25_v_c25 = (F0_test_25d - F0_control_25d),
  F0_t16_v_c16 = (F0_test_16d - F0_control_16d),
  F0_t25_v_t16 = (F0_test_25d - F0_test_16d),
  F0_c25_v_c16 = (F0_control_25d - F0_control_16d),
  F0_t25_v_c16 = (F0_test_25d - F0_control_16d),
  F0_t16_v_c25 = (F0_test_16d - F0_control_25d),
  levels = design_1)

cm_F0

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_F0)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_F0 <- eBayes(fit2_model_1, 
                       # trend = TRUE, #accommodates a mean-variance trend;
                       # robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_F0) %>% summary()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
F0_t25_v_c25 <- topTable(
  results_F0,
  coef=c(1),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

F0_t16_v_c16 <- topTable(
  results_F0,
  coef=c(2),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

F0_t25_v_t16 <- topTable(
  results_F0,
  coef=c(3),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

F0_c25_v_c16 <- topTable(
  results_F0,
  coef=c(4),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

F0_t25_v_c16 <- topTable(
  results_F0,
  coef=c(5),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")
  
F0_t16_v_c25 <- topTable(
  results_F0,
  coef=c(6),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1, cm_F0, results_F0)
```

```{r f1_results}
#####

pData(eset_F1) <- pData(eset_F1) |> droplevels()

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 + label, data = pData(eset_F1))

colnames(design_1) <- levels(pData(eset_F1)$label)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_F1, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_F1 <- makeContrasts(
  F1_t25_v_c25 = (F1_test_25d - F1_control_25d),
  levels = design_1)

cm_F1

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_F1)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_F1 <- eBayes(fit2_model_1, 
                       # trend = TRUE, #accommodates a mean-variance trend;
                       # robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_F1) %>% summary()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
F1_t25_v_c25 <- topTable(
  results_F1,
  coef=c(1),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1, cm_F1, results_F1)
```

```{r f2_results}
#####

pData(eset_F2) <- pData(eset_F2) |> droplevels()

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 + label, data = pData(eset_F2))

colnames(design_1) <- levels(pData(eset_F2)$label)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_F2, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_F2 <- makeContrasts(
  F2_t25_v_c25 = (F2_test_25d - F2_control_25d),
  levels = design_1)

cm_F2

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_F2)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_F2 <- eBayes(fit2_model_1, 
                       # trend = TRUE, #accommodates a mean-variance trend;
                       # robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_F2) %>% summary()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
F2_t25_v_c25 <- topTable(
  results_F2,
  coef=c(1),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1, cm_F2, results_F2)
```

## Report generation
```{r annotation}

library(drosophila2.db) # annotation package for the model organism


#annotation returns mutliple matches for some of the probes so perform it after obtaining DE gene list 
anno_eset <- AnnotationDbi::select(drosophila2.db,
                                       keys = (featureNames(eset_F2)), #indexes the rows
                                       columns = c("SYMBOL", "GENENAME", "ENTREZID"), #indexes the columns
                                       keytype = "PROBEID") %>% #arranges the output by this column
  filter(!is.na(SYMBOL)) %>% # this also removes quality control probes
  group_by(PROBEID) %>%
  filter(n() == 1) # remove  probes which have multiple genes associated with them
# check that all quality control probes, containing the pattern AFFX in their name, were removed
# str_detect(anno_eset$PROBEID, "AFFX") %>% sum()
```

```{r write_files}
#####
library(writexl)

kp_results_list <- vector("list", 8)
names(kp_results_list)  <- c("F0_t25_v_c25","F0_t16_v_c16", "F0_t25_v_t16", "F0_c25_v_c16", "F0_t25_v_c16", "F0_t16_v_c25",  "F1_t25_v_c25", "F2_t25_v_c25") 


kp_results_list[[1]] <- F0_t25_v_c25 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[2]] <- F0_t16_v_c16 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[3]] <- F0_t25_v_t16 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[4]] <- F0_c25_v_c16 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[5]] <- F0_t25_v_c16 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[6]] <- F0_t16_v_c25 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[7]] <- F1_t25_v_c25 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

kp_results_list[[8]] <- F2_t25_v_c25 |> rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% dplyr::rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME, entrez_id = ENTREZID) |> dplyr::select(-c(t, B, AveExpr))

write_xlsx(kp_results_list, path = "C:/Users/abhalla/Documents/microarray/kp_results.xlsx")
```

```{r volcano_plots}
bind_rows("F0" = F0_t25_v_c25, 
          "F1" = F1_t25_v_c25, 
          "F2" = F2_t25_v_c25, 
          .id = "generation") %>% 
  mutate(colo = case_when(P.Value <= 0.05 & abs(logFC) >= 0.585 ~ "1", TRUE ~ "0")) %>% 
  ggplot(aes(x = logFC, y = -log(P.Value, 10), colour = colo)) +
  geom_point(shape= 20) + 
  # xlim(-2,2) +
  ylim(0,5) +
  scale_color_manual(values = c("darkgray", "red")) +
  facet_wrap(~ #scales = "free",
             generation) +
  ylab(label = expression(paste(-Log[10], "(", italic(p), "-value", ")"))) +
  xlab(label = expression(paste(Log[2], "(Fold Change)"))) + 
  theme_classic() +
  theme(legend.position = "none")
```

```{r venn_diagrams}
f0 <- F0_t25_v_c25 %>% rownames_to_column() |> filter(logFC <= 0.05, P.Value <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 

f1 <- F1_t25_v_c25 %>% rownames_to_column() |> filter(logFC <= 0.05, P.Value <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 

f2 <- F2_t25_v_c25 %>% rownames_to_column() |> filter(logFC <= 0.05, P.Value <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 


library(VennDiagram)

# Chart
venn.diagram(
        x = list(f0$rowname, f1$rowname),
        category.names = c("F0" , "F1"),
        filename = 'F0vF1_venn_diagram.png',
        disable.logging=T)

venn.diagram(
        x = list(f1$rowname, f2$rowname),
        category.names = c("F1" , "F2"),
        filename = 'F1vF2_venn_diagram.png',
        disable.logging=T)

venn.diagram(
        x = list(f0$rowname, f2$rowname),
        category.names = c("F0" , "F2"),
        filename = "F0vF2_venn_diagram.png",
        disable.logging=T)

venn.diagram(
        x = list(f0$rowname, f1$rowname, f2$rowname),
        category.names = c("F0" , "F1", "F2"),
        filename = "F0vF1vF2_venn_diagram.png",
        disable.logging=T)
```

```{r deg_correlation}
#correlation between the logFC values of DE genes that are common between different genotype pairs

#function for finding correlation between DEGs found in different pairs
correlate_degs <- function(df1, df2){
  df1 %>% get() %>% filter(P.Value <= 0.05) %>% dplyr::select(rowname, logFC) %>% 
    inner_join(
  df2 %>% get() %>% filter(P.Value <= 0.05) %>% dplyr::select(rowname, logFC), by = "rowname", suffix = c("1", "2")
               ) %>% 
  {cor.test(.$logFC1, .$logFC2)} %>% broom::tidy()
#the curly brackets are needed to suppress the default behavior of the pipe operator, otherwise the data frame to the left of the pipe would have been forwarded as the third argument into cor(), which only takes two arguments.
}

#run the correlate_degs() function on all the dfs in the environment
bind_rows(
  map2_df(.x = "f0", 
        .y = "f1",
        .f = correlate_degs),
  map2_df(.x = "f1",
        .y = "f2",
        .f = correlate_degs),
    .id = "id"
  )

#function for finding DEGs that overlap within different pairs
join_degs <- function(df1, df2){
  df1 %>% get() %>% filter(P.Value <= 0.05) %>% dplyr::select(rowname, logFC) %>% 
    inner_join(
  df2 %>% get() %>% filter(P.Value <= 0.05) %>% dplyr::select(rowname, logFC),
  by = "rowname", suffix = c("1", "2")
    ) %>% mutate(comparison = 
                   paste(rlang::expr_name(df1)|> toupper(), 
                         "vs", 
                         rlang::expr_name(df2)|> toupper(), 
                         sep = " ")
                 ) # adds the name of the df2 into the group2 column
  }

#run the  join_degs() function on all the dfs in the environment
map2_df(.x = "f0", .y = "f0", .f = join_degs) %>% 
  bind_rows(
map2_df(.x = "f0", .y = "f1", .f = join_degs)) %>% 
  bind_rows(
map2_df(.x = "f1", .y = "f2", .f = join_degs)) %>%   
  ggplot(aes(logFC1, logFC2)) + 
  geom_point(
    alpha = 0.20,
    shape = 1
    ) + 
  geom_line(stat = "smooth", method = "lm", formula = 'y ~ x', alpha = 0.5) +
  facet_wrap(~ comparison, scales = "fixed") +
  theme_bw() +
  theme(aspect.ratio = 1) +
  ggtitle("Most fold changes negative! Why?")
```

```{r GO}

library(clusterProfiler)

f0 |> dplyr::pull(ENTREZID) |> enrichGO(OrgDb = "drosophila2.db", keyType = "ENTREZID", ont = "BP")  |> 
  barplot(title = "Biological Processes Enriched Among DE Genes in F0")
f1 |> dplyr::pull(ENTREZID) |> enrichGO(OrgDb = "drosophila2.db", keyType = "ENTREZID", ont = "BP")  |> 
    barplot(title = "Biological Processes Enriched Among DE Genes in F1")
f2 |> dplyr::pull(ENTREZID) |> enrichGO(OrgDb = "drosophila2.db", keyType = "ENTREZID", ont = "BP")  |>
    barplot(title = "Biological Processes Enriched Among DE Genes in F2")


# # GO common between f0 & f1 generations
# join_degs("f0", "f1") |> left_join(anno_eset, by = c("rowname" = "PROBEID")) |> dplyr::pull(ENTREZID) |> enrichGO(OrgDb = "drosophila2.db", keyType = "ENTREZID", ont = "BP")  |> barplot()
# 
# # GO common between f1 & f2 generations: returns error
# join_degs("f1", "f2") |> left_join(anno_eset, by = c("rowname" = "PROBEID")) |> dplyr::pull(ENTREZID) |> enrichGO(OrgDb = "drosophila2.db", keyType = "ENTREZID", ont = "BP")  |> barplot()
```

```{r GEO_files}
matrix_F0 <- eset_F0 |> exprs() |> as.data.frame() |> rownames_to_column() |> dplyr::rename(ID_REF = rowname) 
matrix_F1 <- eset_F1 |> exprs() |> as.data.frame() |> rownames_to_column() |> dplyr::rename(ID_REF = rowname) 
matrix_F2 <- eset_F2 |> exprs() |> as.data.frame() |> rownames_to_column() |> dplyr::rename(ID_REF = rowname)

left_join(matrix_F0, matrix_F1, by = "ID_REF") |> left_join(matrix_F2, by = "ID_REF") |> write_csv( file = "C:/Users/abhalla/Documents/microarray/matrix_all.csv")

kp_samples |> mutate("Sample name" = paste0(label, "_", replicate)) |> write_csv(file = "C:/Users/abhalla/Documents/microarray/metadata.csv")
```
