---
title: "Microarray Pipeline"
author: "Ameek Bhalla"
editor_options:
  chunk_output_type: console
output:
  html_notebook:
    number_sections: yes
    theme: readable
    toc: yes
    toc_depth: '2'
    toc_float: yes
  html_document:
    df_print: paged
    number_sections: yes
    theme: readable
    toc: yes
    toc_depth: '2'
    toc_float: yes
bibliography: Microarray references.bib
---

***
# Setup
The default chunk options can be set in the first chunk itself. Later, the behaviour of individuals chunks can be modified by modifying a specific argument.

```{r setup results = "hide"}
## Set default chunk options

knitr::opts_chunk$set(results = "hide", 
                      tidy = TRUE)
## Load Packages
library(affy) # Bioconductor package for analysis of older Affymetrix arrays
#library(affyio)
#library(affyPLM)
#library(annaffy)
#library(annotate)
#library(arrayQualityMetrics)
library(BiocManager)
library(drosophila2.db) # annotation package for the model organism
library(drosophila2cdf)
library(formatR)
library(gcrma)
library(htmltools)
library(knitr)
library(limma)
#library(markdown)
library(rmarkdown)
#library(simpleaffy)
library(tidyverse)
```

# Importing data into R
```{r get_data}

sample_data <- read_csv(
  "sample_data.csv", 
  col_types = cols(
    age = col_integer(), 
    day = col_factor(levels = c("1", "2", "3", "4", "5", "6")), 
    generation = col_factor(levels = c("F-1", "F0 + F1", "F2", "F3")), 
    module = col_factor(levels = c("1", "2", "3", "4")), 
    number = col_integer(), 
    replicate = col_factor(levels = c("L3", "L4", "L5", "L7", "L8", "L9", "L10")), 
    sex = col_factor(levels = c("female", "male")), 
    stored = col_date(format = "%m/%d/%Y"), 
    wash = col_factor(levels = c("1", "2", "3", "4")),
    genotype = col_factor(levels = c(
      "control",  "test", "w1118", "rpr/cyo; dilp3/ser", "cyo/+; ser/+", "rpr/+; dilp3/+", "cyo/+; dilp3/+", "rpr/+; ser/+"))))


raw_f1_f <- sample_data %>% filter(generation == "F0 + F1", sex == "female") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)

raw_f2_f <- sample_data %>% filter(generation == "F2", sex == "female") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)

raw_f3_f <- sample_data %>% filter(generation == "F3", sex == "female") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)


raw_f1_m <- sample_data %>% filter(generation == "F0 + F1", sex == "male") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)

raw_f2_m <- sample_data %>% filter(generation == "F2", sex == "male") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)

raw_f3_m <- sample_data %>% filter(generation == "F3", sex == "male") %>%
  ReadAffy(filenames = .$file_name,
           celfile.path = "./CEL Files",
           sampleNames = .$label,
           phenoData = .)

eset_f1_f <- gcrma(raw_f1_f)
eset_f2_f <- gcrma(raw_f2_f)
eset_f3_f <- gcrma(raw_f3_f)

eset_f1_m <- gcrma(raw_f1_m)
eset_f2_m <- gcrma(raw_f2_m)
eset_f3_m <- gcrma(raw_f3_m)
```

```{r plot_intensities eval = FALSE}
raw_expression_f1_f <- exprs(raw_f1_f) %>% #plotDensities() accepts only matrix/ExpressionSet objects
                  log2() #the data is log2 scaled to linearize it across different orders of magnitude 
raw_expression_f2_f <- exprs(raw_f2_f) %>% #plotDensities() accepts only matrix/ExpressionSet objects
                  log2() #the data is log2 scaled to linearize it across different orders of magnitude 


plotDensities(raw_expression_f1_f, group = pData(raw_f1_f)$genotype, legend = "topright",
              main = "Variation in expression according to genotype (raw_f1_f)")
plotDensities(raw_expression_f2_f, group = pData(raw_f2_f)$genotype, legend = "topright",
              main = "Variation in expression according to genotype (raw_f2_f)")


plotDensities(raw_expression_f1_f, group = pData(raw_f1_f)$replicate, legend = "topright",
              main = "Variation in expression according to replicate (raw_f1_f)")
plotDensities(raw_expression_f2_f, group = pData(raw_f2_f)$replicate, legend = "topright",
              main = "Variation in expression according to replicate (raw_f2_f)")




norm_expression_f1_f <- exprs(eset_f1_f) %>% #plotDensities() accepts only matrix/ExpressionSet objects
                  log2() #the data is log2 scaled to linearize it across different orders of magnitude 
norm_expression_f2_f <- exprs(eset_f2_f) %>% #plotDensities() accepts only matrix/ExpressionSet objects
                  log2() #the data is log2 scaled to linearize it across different orders of magnitude 

plotDensities(norm_expression_f1_f, group = pData(eset_f1_f)$genotype, legend = "topright",
              main = "Variation in expression according to genotype (eset_f1_f)")
plotDensities(norm_expression_f2_f, group = pData(eset_f2_f)$genotype, legend = "topright",
              main = "Variation in expression according to genotype (eset_f2_f)")

plotDensities(norm_expression_f1_f, group = pData(eset_f1_f)$replicate, legend = "topright",
              main = "Variation in expression according to replicate (eset_f1_f)")
plotDensities(norm_expression_f2_f, group = pData(eset_f2_f)$replicate, legend = "topright",
              main = "Variation in expression according to replicate (eset_f2_f)")
```

# Differential expression analysis
## Model formula 1: Group-means parametrization 
```{r group_means_interaction}
#####
pData(eset_f1_f) <- pData(eset_f1_f) %>% 
  mutate(
    gtype_clean = case_when( # categorize genotypes into test or control lineages
      genotype == "rpr/cyo; dilp3/ser" ~ "rpr.cyo_dilp3.ser",
      genotype == "rpr/+; dilp3/+" ~ "rpr_dilp3",
      genotype == "cyo/+; dilp3/+" ~ "cyo_dilp3",
      genotype == "rpr/+; ser/+" ~ "rpr_ser",
      genotype == "cyo/+; ser/+" ~ "cyo_ser",
      TRUE ~ "w1118") %>% as_factor())


# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + gtype_clean, data = pData(eset_f1_f))

colnames(design_1) <- levels(pData(eset_f1_f)$gtype_clean)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f1_f, design_1, block =  pData(eset_f1_f)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_mini, design_1, 
#                      block =  pData(eset)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f1_f, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f1_f <- makeContrasts(
  rpr.cyo_dilp3.ser = (rpr.cyo_dilp3.ser - w1118),
  cyo_ser = (cyo_ser - w1118),
  rpr_dilp3 = (rpr_dilp3 - w1118),
  cyo_dilp3 = (cyo_dilp3 - w1118),
  rpr_ser = (rpr_ser  - w1118),
  levels = design_1)

cm_f1_f

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f1_f)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f1_f <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f1_f) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f1_f) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
rpr.cyo_dilp3.ser_f <- topTable(
  results_f1_f,
  coef=c(1),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

cyo_ser_f <- topTable(
  results_f1_f,
  coef=c(2),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

rpr_dilp3_f <- topTable(
  results_f1_f,
  coef=c(3),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

cyo_dilp3_f <- topTable(
  results_f1_f,
  coef=c(4),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

rpr_ser_f <- topTable(
  results_f1_f,
  coef=c(5),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

#####
pData(eset_f1_m) <- pData(eset_f1_m) %>% 
  mutate(
    gtype_clean = case_when( # categorize genotypes into test or control lineages
      genotype == "rpr/cyo; dilp3/ser" ~ "rpr.cyo_dilp3.ser",
      genotype == "rpr/+; dilp3/+" ~ "rpr_dilp3",
      genotype == "cyo/+; dilp3/+" ~ "cyo_dilp3",
      genotype == "rpr/+; ser/+" ~ "rpr_ser",
      genotype == "cyo/+; ser/+" ~ "cyo_ser",
      TRUE ~ "w1118") %>% as_factor())


# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + gtype_clean, data = pData(eset_f1_m))

colnames(design_1) <- levels(pData(eset_f1_m)$gtype_clean)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f1_m, design_1, block =  pData(eset_f1_m)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_mini, design_1, 
#                      block =  pData(eset)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f1_m, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f1_m <- makeContrasts(
  rpr.cyo_dilp3.ser = (rpr.cyo_dilp3.ser - w1118),
  cyo_ser = (cyo_ser - w1118),
  rpr_dilp3 = (rpr_dilp3 - w1118),
  cyo_dilp3 = (cyo_dilp3 - w1118),
  rpr_ser = (rpr_ser  - w1118),
  levels = design_1)

cm_f1_m

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f1_m)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f1_m <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f1_m) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f1_m) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
rpr.cyo_dilp3.ser_m <- topTable(
  results_f1_m,
  coef=c(1),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

cyo_ser_m <- topTable(
  results_f1_m,
  coef=c(2),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

rpr_dilp3_m <- topTable(
  results_f1_m,
  coef=c(3),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

cyo_dilp3_m <- topTable(
  results_f1_m,
  coef=c(4),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

rpr_ser_m <- topTable(
  results_f1_m,
  coef=c(5),
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

hist(rpr.cyo_dilp3.ser_m$P.Value)
hist(cyo_ser_m$P.Value)
hist(rpr_dilp3_m$P.Value)
hist(cyo_dilp3_m$P.Value)
hist(rpr_ser_m$P.Value)
#####
group <- with(pData(eset_f2_f), 
              paste(generation,
                    genotype, 
                    # sex,
                    # day,
                    sep = "."))

pData(eset_f2_f)$group <- as.factor(group)
remove(group)

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + group, data = pData(eset_f2_f))
colnames(design_1) <- levels(pData(eset_f2_f)$group)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f2_f, design_1, block =  pData(eset_f2_f)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_f2_f, design_1, 
#                      block =  pData(eset_f2_f)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f2_f, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f2_f <- makeContrasts(
  f2_f = (F2.test - F2.control),
  levels = design_1)

cm_f2_f

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f2_f)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f2_f <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f2_f) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f2_f) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
f2_f <- topTable(
  results_f2_f,
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

hist(f2_f$P.Value)
#####
group <- with(pData(eset_f2_m), 
              paste(generation,
                    genotype, 
                    # sex,
                    # day,
                    sep = "."))

pData(eset_f2_m)$group <- as.factor(group)
remove(group)

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + group, data = pData(eset_f2_m))
colnames(design_1) <- levels(pData(eset_f2_m)$group)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f2_m, design_1, block =  pData(eset_f2_m)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_f2_m, design_1, 
#                      block =  pData(eset_f2_m)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f2_m, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f2_m <- makeContrasts(
  f2_m = (F2.test - F2.control),
  levels = design_1)

cm_f2_m

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f2_m)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f2_m <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f2_m) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f2_m) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
f2_m <- topTable(
  results_f2_m,
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

hist(f2_m$P.Value)
#####
group <- with(pData(eset_f3_f), 
              paste(generation,
                    genotype, 
                    # sex,
                    # day,
                    sep = "."))

pData(eset_f3_f)$group <- as.factor(group)
remove(group)

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + group, data = pData(eset_f3_f))
colnames(design_1) <- levels(pData(eset_f3_f)$group)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f3_f, design_1, block =  pData(eset_f3_f)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_f3_f, design_1, 
#                      block =  pData(eset_f3_f)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f3_f, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f3_f <- makeContrasts(
  f3_f = (F3.test - F3.control),
  levels = design_1)

cm_f3_f

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f3_f)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f3_f <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f3_f) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f3_f) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
f3_f <- topTable(
  results_f3_f,
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

hist(f3_f$P.Value)
#####
group <- with(pData(eset_f3_m), 
              paste(generation,
                    genotype, 
                    # sex,
                    # day,
                    sep = "."))

pData(eset_f3_m)$group <- as.factor(group)
remove(group)

# Create design matrix with group-means parametrization (no intercept)
design_1 <- model.matrix(~ 0 
                         # + day 
                         + group, data = pData(eset_f3_m))
colnames(design_1) <- levels(pData(eset_f3_m)$group)
colnames(design_1)

# Count the number of samples modeled by each coefficient
colSums(design_1)

#estimate the correlation between measurements made on the same repliactes:
# corfit_1 <- duplicateCorrelation(eset_f3_m, design_1, block =  pData(eset_f3_m)$replicate)
# corfit_1$consensus

# corfit_1$consensus <- by_sex$both[[1]]$consensus.correlation
# corfit_1$cor <- by_sex$both[[1]]$cor
# corfit_1$atanh.correlations <- by_sex$both[[1]]$atanh.correlations

#fit the linear model fit_model_1 with blocking
# fit_model_1 <- lmFit(eset_f3_m, design_1, 
#                      block =  pData(eset_f3_m)$replicate, 
#                      correlation = corfit_1$consensus)

# #fit the linear model fit_model_1 without blocking
fit_model_1 <- lmFit(eset_f3_m, design_1)

## Build the contrasts matrix for discovering main effect of genotype in females

cm_f3_m <- makeContrasts(
  f3_m = (F3.test - F3.control),
  levels = design_1)

cm_f3_m

#fit the contrast matrix
fit2_model_1 <- contrasts.fit(fit_model_1, cm_f3_m)

#set trend = TRUE  since the data do not show a uniform variance across all expression intensities (this trend in the variance at different mean intensities is discovered with `plotSA(fit_model)` ) 
results_f3_m <- eBayes(fit2_model_1, 
                       trend = TRUE, #accommodates a mean-variance trend; 
                       robust = TRUE #increases power for some kinds of studies
                       ) 

#Results of all contrasts simultaneously
decideTests(results_f3_m) %>% summary()

# #Venn diagram showing numbers of genes significant in all contrasts, simultaneously
decideTests(results_f3_m) %>% vennDiagram()

#A list of top genes differentially expressed in a contarst can be obtained by specifying its coef
# topTable(fit2_model_1, coef=c(1:3), adjust="BH", number = 30)
f3_m <- topTable(
  results_f3_m,
  adjust="BH", number = nrow(fit2_model_1), sort.by = "none")

remove(design_1, fit_model_1, fit2_model_1)

hist(f3_m$P.Value)
```

## Empirical null distributions
```{r annotation}
#annotation returns mutliple matches for some of the probes so perform it after obtaining DE gene list 
anno_eset <- AnnotationDbi::select(drosophila2.db,
                                       keys = (featureNames(eset_f3_m)), #indexes the rows
                                       columns = c("SYMBOL", "GENENAME"), #indexes the columns
                                       keytype = "PROBEID") %>% #arranges the output by this column
  filter(!is.na(SYMBOL)) %>% # this also removes quality control probes
  group_by(PROBEID) %>%
  filter(n() == 1) # remove  probes which have multiple genes associated with them
# check that all quality control probes, containing the pattern AFFX in their name, were removed
# str_detect(anno_eset$PROBEID, "AFFX") %>% sum()
```

```{r empirical_null}

library(fdrtool)

#####
rpr.cyo_dilp3.ser_f <- (rpr.cyo_dilp3.ser_f$t - median(rpr.cyo_dilp3.ser_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr.cyo_dilp3.ser_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

cyo_ser_f <- (cyo_ser_f$t - median(cyo_ser_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(cyo_ser_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

rpr_dilp3_f <- (rpr_dilp3_f$t - median(rpr_dilp3_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr_dilp3_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

cyo_dilp3_f <- (cyo_dilp3_f$t - median(cyo_dilp3_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(cyo_dilp3_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

rpr_ser_f <- (rpr_ser_f$t - median(rpr_ser_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr_ser_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

#####
rpr.cyo_dilp3.ser_m <- (rpr.cyo_dilp3.ser_m$t - median(rpr.cyo_dilp3.ser_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr.cyo_dilp3.ser_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

cyo_ser_m <- (cyo_ser_m$t - median(cyo_ser_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(cyo_ser_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

rpr_dilp3_m <- (rpr_dilp3_m$t - median(rpr_dilp3_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr_dilp3_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

cyo_dilp3_m <- (cyo_dilp3_m$t - median(cyo_dilp3_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(cyo_dilp3_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()

rpr_ser_m <- (rpr_ser_m$t - median(rpr_ser_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(rpr_ser_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()
#####
f2_f <- (f2_f$t - median(f2_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(f2_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()
#####
f2_m <- (f2_m$t - median(f2_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(f2_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()
#####
f3_f <- (f3_f$t - median(f3_f$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(f3_f, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()
#####
f3_m <- (f3_m$t - median(f3_m$t)) %>% fdrtool(statistic = "normal", plot = F) %>% cbind(f3_m, .) %>% rownames_to_column() %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) %>% rename(probe_id = rowname, symbol = SYMBOL, gene_name = GENENAME) %>% select("probe_id", "symbol", "gene_name", "logFC", "pval", "qval") #  %>% drop_na()
```

```{r write_files}
#####
library(XLConnect)

# by_generation2 <- loadWorkbook("by_generation2.xlsx", create = TRUE)
# 
# createSheet(by_generation2, "rpr.cyo_dilp3.ser_f0_f")
# writeWorksheet(by_generation2, rpr.cyo_dilp3.ser_f, "rpr.cyo_dilp3.ser_f0_f")
# saveWorkbook(by_generation2, "by_generation2.xlsx")
# 
# createSheet(by_generation2, "cyo_ser_f1_f")
# writeWorksheet(by_generation2, cyo_ser_f, "cyo_ser_f1_f")
# saveWorkbook(by_generation2, "by_generation2.xlsx")
# 
# createSheet(by_generation2, "rpr_dilp3_f1_f")
# writeWorksheet(by_generation2, rpr_dilp3_f, "rpr_dilp3_f1_f")
# saveWorkbook(by_generation2, "by_generation2.xlsx")

by_generation2.2 <- loadWorkbook("by_generation2.2.xlsx", create = TRUE)
createSheet(by_generation2.2, "cyo_dilp3_f1_f")
writeWorksheet(by_generation2.2, cyo_dilp3_f, "cyo_dilp3_f1_f")
saveWorkbook(by_generation2.2, "by_generation2.2.xlsx")

by_generation2.3 <- loadWorkbook("by_generation2.3.xlsx", create = TRUE)
createSheet(by_generation2.3, "rpr_ser_f1_f")
writeWorksheet(by_generation2.3, rpr_ser_f, "rpr_ser_f1_f")
saveWorkbook(by_generation2.3, "by_generation2.3.xlsx")

by_generation2.4 <- loadWorkbook("by_generation2.4.xlsx", create = TRUE)
createSheet(by_generation2.4, "f2_f")
writeWorksheet(by_generation2.4, f2_f, "f2_f")
saveWorkbook(by_generation2.4, "by_generation2.4.xlsx")

# by_generation2.5 <- loadWorkbook("by_generation2.5.xlsx", create = TRUE)
# createSheet(by_generation2.5, "f3_f")
# writeWorksheet(by_generation2.5, f3_f, "f3_f")
# saveWorkbook(by_generation2.5, "by_generation2.5.xlsx")
```

```{r counts}
by_generation <- NULL
by_generation <- by_generation %>% as_tibble()

by_generation[1,1] <- rpr.cyo_dilp3.ser_f %>% filter(pval <= 0.05) %>% count()
by_generation[2,1] <- cyo_ser_f %>% filter(pval <= 0.05) %>% count()
by_generation[3,1] <- rpr_dilp3_f %>% filter(pval <= 0.05) %>% count()
by_generation[4,1] <- cyo_dilp3_f %>% filter(pval <= 0.05) %>% count()
by_generation[5,1] <- rpr_ser_f %>% filter(pval <= 0.05) %>% count()
by_generation[6,1] <- f2_f %>% filter(pval <= 0.05) %>% count()
by_generation[7,1] <- f3_f %>% filter(pval <= 0.05) %>% count()

by_generation <- by_generation %>% rename(females_unblocked = n)
by_generation$data <- c(
  "rpr.cyo_dilp3.ser_f",
  "cyo_ser_f",
  "rpr_dilp3_f",
  "cyo_dilp3_f",
  "rpr_ser_f", 
  "f2_f", 
  "f3_f"
)

```

```{r volcano_plots}
bind_rows("F0" = geno_0 %>% left_join(anno_eset, by = c("rowname" = "PROBEID")), 
          "F1" = geno_1 %>% left_join(anno_eset, by = c("rowname" = "PROBEID")), 
          "F2" = geno_2 %>% left_join(anno_eset, by = c("rowname" = "PROBEID")), 
          "F3" = geno_3 %>% left_join(anno_eset, by = c("rowname" = "PROBEID")), 
          .id = "generation") %>% 
  mutate(colo = case_when(pval <= 0.05 & abs(logFC) >= 0.585 ~ "1", TRUE ~ "0")) %>% 
  ggplot(aes(x = logFC, y = -log(pval, 10), colour = colo)) +
  geom_point(shape= 20) + 
  # xlim(-2,2) +
  ylim(0,15) +
  scale_color_manual(values = c("darkgray", "red")) +
  facet_wrap(~ #scales = "free",
             generation) +
  theme_classic() +
  theme(legend.position = "none")
```

```{r venn_diagrams}
f0_f <- geno_0 %>% filter(logFC <= 0.05, pval <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 
f1_f <- geno_1 %>% filter(logFC <= 0.05, pval <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 

f0_m <- geno_0 %>% filter(logFC <= 0.05, pval <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 
f1_m <- geno_1 %>% filter(logFC <= 0.05, pval <= 0.05) %>% left_join(anno_eset, by = c("rowname" = "PROBEID")) 


library(VennDiagram)

# Chart
venn.diagram(
        x = list(f1_f$rowname, f1_m$rowname),
        category.names = c("F1 Female" , "F1 Male"),
        filename = 'F1_venn_diagramm.png',
        output=TRUE)

```